### PR TITLE
[DoNotMerge] [CompilerPerf] All your collections are belong to us

### PR DESCRIPTION
**Pull request by manofstick**
This PR will be continually rebased off #2745.

_It is **not** expected to ever be pulled into the main line, but rather this is just testing conformance and performance of Seq by changing the List.* and Array.* functions to Seq functions within the compiler (and beyond maybe.)_

### CODE REVIEW

File: src/absil/il.fs [hunk 1]
```diff
@@ -165,7 +165,7 @@ type LazyOrderedMultiMap<'Key,'Data when 'Key : equality>(keyf : 'Data -> 'Key,
     let quickMap= 
         lazyItems |> lazyMap (fun entries -> 
             let t = new Dictionary<_,_>(entries.Length, HashIdentity.Structural)
-            do entries |> List.iter (fun y -> let key = keyf y in t.[key] <- y :: (if t.ContainsKey(key) then t.[key] else [])) 
+            do entries |> Seq.iter (fun y -> let key = keyf y in t.[key] <- y :: (if t.ContainsKey(key) then t.[key] else []))
```
Comment (forki) [inline src/absil/il.fs]: Is this faster in any way?

Comment (manofstick) [inline src/absil/il.fs]: I severely doubt it :-) This is more a process of checking that the new Seq stuff works on a "real" code base. But I'm interested also in general performance as well - we'll see how it goes!

**Comment by manofstick** [comment]
@dsyme 

I'm getting wild variations in times of the build server; but when I kick off this at about the same time as #2587, I appear to be getting faster builds with this branch. This wasn't really to be expected, as this was more just a test that the Seq functions were not too much worse that directly using the List functions, but in the process I am making *some* things lazy that previously weren't, so it is possible that this may be faster? (i.e. doesn't have to iterate all lists completely, plus saving on garbage)

Anyway, might all be just random luck on the build server - do you have any more consistent figures on this builds timings?

(Even if it isn't, I'll continue to convert more of it over to Seq; as it will help if I do do any final optimizations in ISeq land...)

**Comment by dsyme** [comment]
> I'm getting wild variations in times of the build server; but when I kick off this at about the same time as #2587, I appear to be getting faster builds with this branch. This wasn't really to be expected, as this was more just a test that the Seq functions were not too much worse that directly using the List functions, but in the process I am making some things lazy that previously weren't, so it is possible that this may be faster? (i.e. doesn't have to iterate all lists completely, plus saving on garbage)

I think it's better to keep the use of Seq in the compiler  as a separate branch (on top of this one) so we can measure that independently?  I'd hate there to be some problem where there is either a performance problem or semantic change and we have to dig back through this PR to find it

**Comment by manofstick** [comment]
@dsyme 

This already is a seperate branch 😀 I have three on the go. This is just a test branch really...

**Comment by dsyme** [comment]
> This already is a seperate branch 😀 I have three on the go. This is just a test branch really...

Oh, great, yes, I hadn't noticed.  Thanks for doing it this way.

**Comment by manofstick** [comment]
Don't think I will continue with this experiment. I have just been modifying as many things to seq as I can without too much effort just to make sure that it still works OK. I wasn't really expecting this to be any more performant that it previously was, I wasn't being particularly targeted in my approach, just going to whatever I could do en masse, but this meant that in many (most?) cases all I was doing was removing an existing optimized path for one that needs to do casting and creating more objects. But I am interested to see what the actual performance vs original was so if those figures could be produced at some stage then I'd be happy...

And I'll leave this around for a little while, as I might do some lower level work and see if any performance changes ripple up, or if I damage performance...

**Comment by manofstick** [comment]
Bit of a mess rebasing here; but we'll see if it survives... Not that I care *too* much about this branch, but nice if it remains green... we'll see...

**Comment by manofstick** [comment]
This was just a temporary PR for testing, so closing.

**Pull request closed without merging**